xgcm/grid_ufunc.py

```diff
@@ -335,30 +362,83 @@ def apply_as_grid_ufunc(
             fill_value=fill_value,
         )
+    else:
+        boundary_width_real_axes = {
+            real_ax: (0, 0) for real_ax in dummy_to_real_axes_mapping.values()
+        }
+        padded_args = args
+
+    if (
+        any(
+            _has_chunked_core_dims(arg, core_dims)
+            for arg, core_dims in zip(args, in_core_dims)
+        )
+        and map_overlap
+    ):
+        # map operation over dask chunks along core dimensions
+        from dask.array import map_overlap
+
+        # merge any lonely chunks from padding
+        rechunked_padded_args = _rechunk_to_merge_in_boundary_chunks(
+            padded_args,
+            args,
+            boundary_width_real_axes,
+            grid,
+        )
+
+        # TODO should this be adjusted?
+        boundary_width_per_numpy_axis = {
+            grid.axes[ax_name]._get_axis_dim_num(args[0]): width
+            for ax_name, width in boundary_width_real_axes.items()
+        }
+
+        def _dict_to_numbered_axes(sizes):
+            return tuple(sizes[dim] for dim in sizes.keys())
+
+        output_chunk_length_adjustment = _check_if_length_would_change(  # noqa
+            out_dummy_ax_names, in_ax_pos, out_ax_pos
+        )
+
+        # Output chunks are the same as input chunks (as we disallowed axis positions for which this is not the case)
+        # TODO first argument only because map_overlap can't handle multiple return values (I think)
+        original_chunksizes = [arg.chunksizes for arg in args]
+        true_chunksizes = original_chunksizes[0]
+        true_chunksizes_per_numpy_axis = _dict_to_numbered_axes(true_chunksizes)
+
+        # (we don't need a separate code path using bare map_blocks if boundary_widths are zero because map_overlap just
+        # calls map_blocks automatically in that scenario)
+        def mapped_func(*a, **kw):
+            return map_overlap(
+                func,
+                *a,
+                **kw,
+                depth=boundary_width_per_numpy_axis,
+                boundary="none",
+                trim=False,
+                meta=np.array([], dtype=out_dtypes[0]),
+                chunks=true_chunksizes_per_numpy_axis,
+            )
+
+    else:
+        rechunked_padded_args = padded_args
+        mapped_func = func
 
     # Determine expected output dimension sizes from grid._ds
     # Only required when dask='parallelized'
-    # TODO does padding change this?
     out_sizes = {
         out_dim: grid._ds.dims[out_dim] for arg in out_core_dims for out_dim in arg
     }
-    # TODO allow users to specify new output dtypes
-    out_dtypes = [a.dtype for a in args]
-
-    # TODO Map operation over dask chunks?
-    # def mapped_func(*a, **kw):
-    #    return map_overlap(func, *a, **kw, depths=boundary_depths, boundary=None, trim=True)
 
     # Perform operation via xarray.apply_ufunc
     results = xr.apply_ufunc(
-        func,
-        *args,
+        mapped_func,
+        *rechunked_padded_args,
```

Comment:
They are the args, exclusively. We can't easily support kwargs as xarray data objects because
a) IIRC `apply_ufunc` only does alignment and so on to xarray objects that are passed as args, any kwargs passed to `apply_ufunc` are just immediately passed on to the wrapped numpy ufunc,
b) How would our signature specify which kwarg was in which grid position? We would have to change the format at least.

I'll try and make this clearer in the docstrings.